spec/ddtrace/contrib/sucker_punch/patcher_spec.rb

```diff
@@ -9,12 +9,17 @@
       c.use :sucker_punch
     end
 
-    SuckerPunch::Queue.clear
     SuckerPunch::RUNNING.make_true
   end
 
   after do
+    count = Thread.list.size
+
+    SuckerPunch::RUNNING.make_false
+    SuckerPunch::Queue.all.each(&:shutdown)
     SuckerPunch::Queue.clear
+
+    try_wait_until { Thread.list.size < count }
   end
```

Comment:
I suspect we can use [`wait_for_termination`](http://ruby-concurrency.github.io/concurrent-ruby/1.1.8/Concurrent/ThreadPoolExecutor.html) to make this a bit cleaner 

```ruby
  after do
    SuckerPunch::RUNNING.make_false
    queues = SuckerPunch::Queue.all
    queues.each(&:shutdown)
    SuckerPunch::Queue.clear
    queues.each { |it| raise 'timeout...' unless it.wait_for_termination(5) }
  end
```

Reply:
I gave this suggestion a try, but it `wait_for_termination` only waits for the request for termination to be received, not for all threads to be terminated.

The thread that is "leaking" here [is this one](https://github.com/ruby-concurrency/concurrent-ruby/blob/a5076bf9e977b290cce5a65b7b72810562bca295/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb#L310).

And unfortunately, the interface to interact with it is to [`#kill` it](https://github.com/ruby-concurrency/concurrent-ruby/blob/a5076bf9e977b290cce5a65b7b72810562bca295/lib/concurrent-ruby/concurrent/executor/ruby_thread_pool_executor.rb#L325-L327) which does not wait for a complete shutdown.

Overall the ThreadPoolExecutor doesn't have a strong interface for waiting for complete termination. This is something we can follow up with the concurrent-ruby team on.

Reply:
That's a shame. Thanks for taking a stab at it. Maybe it would make sense to leave here a comment mentioning that we didn't find a nicer way because of the issue you mentioned?